fontselect: use language from track for fallback selection

rcombs opened the pull request:
Closes #550.

We'll want to encourage players to set `track->Language` to any value they might have from file metadata. Demo mpv patch: https://gist.github.com/1619f3a49ee72b08aa4428b2a7cfa70c.

This doesn't remove the existing code that removes the default FC_LANG value on Fontconfig, as I'm still not entirely sure what cases that was referencing having issues with.

TheOneric commented:
Currently the `Language` Script Info header only accepts two-letter ISO 3166-1 codes. Any extension of the supported notation should imo also be made for the in-file header (which is a libass-specific extension) not just for manual struct assignments.

Something to consider is that different libraries we interact with support different language notations. Your commit mentions UTS#35, HarfBuzz appears to use IETF BCP 47 codes as noted in [#404](https://github.com/libass/libass/issues/404#issuecomment-648306753) and libunibreak can only handle ISO 3166-1 iirc.
I don't know how compatible UTS#35 and IETF BCP 47 are, at first glance they appear to have at least some similarities.
If this PR lands before #486 it doesn't need to consider the only-ISO-3166-1 case much, but we should recommend to prefer two-letters codes over three-letter codes or numeric notations when available. If it lands after #486 the language tag must be sanitised before being passed to libunibreak.

astiob commented on libass/ass_fontselect.h:
```diff
@@ -148,13 +148,15 @@ typedef void    (*SubstituteFontFunc)(void *priv, const char *name,
  * \param lib ASS_Library instance
  * \param family original font family name (try matching a similar font) (never NULL)
  * \param codepoint Unicode codepoint (UTF-32)
+ * \param locale UTS#35 locale to use for the lookup, or NULL for system default
```
Minor: the mention of UTS#35 seems awfully specific. Do we know that our font provider backends specifically expect _Unicode locale identifiers?_ Does HarfBuzz (which is what the tracks’s `Language` is currently used for) expect the same? And at any rate, we only currently read two bytes of the `Language` header (but technically a consumer may put arbitrary data in through the API), although IIRC we’ve had requests to fix—er, expand this.

rcombs replied:
UTS#35 is what CoreText expects; Fontconfig takes RFC3066 in a way that makes both forms effectively equivalent. We could call it "RFC 3066" here if we wanted since that's technically slightly stricter.

rcombs replied:
("BCP 47" is another common name for these)

TheOneric replied:
Fontconfig according to its documentation accepts “the form Ll-Tt where Ll is a two or three letter language from ISO 639 and Tt is a territory from ISO 3166”.

Notably, this differs from BCP47 in that the latter can have a “script” identifier before the region identifier which Fontconfig doesn't expect or understand and in that BCP47 requires the shortest ISO 369 language code to be used while Fontconfig doesn't.

If we take BCP47 then without maintaining our own database of identifiers, we can iinm pass it as is to HarfBuzz, check if the first part consists of two letters and only pass it to libunibreak if that’s the case. It appears like script tags are always 4letter while region tags can have 2 or 3 letters. We could use this to sanitise the identifier to a form accepted by Fontconfig. Or can `FcLangNormalize` already handle this?

Nit: IETF BCP 47 actually contains RFC5646, which obsoleted the RFC obsoleting RFC3066.

TheOneric replied:
> It appears like script tags are always 4letter while region tags can have 2 or 3 letters. We could use this to sanitise the identifier to a form accepted by Fontconfig. Or can FcLangNormalize already handle this?

With the documented restrictions on which BCP-47 subtag we take this _should_ not be a problem anymore and just work if every API users follows docs

astiob reviewed (commented):
One clear bug aside, looks okayish.

I’m worried by the linearly searched, quadratically reallocated array in `ass_fontconfig`, but then one would hope this should be irrelevant. But as an idea, how about making the `ProviderLocale`s form a singly-linked list? That would avoid the quadratic reallocation and possibly look cleaner (as `nb_locales` and `ASS_REALLOC_ARRAY` would be unnecessary), while of course maintaining the linear search.

Do I understand correctly that `locale` is needed in `ass_font_desc` because a single font cache may be used with multiple `ASS_Track`s?

rcombs commented on libass/ass_fontconfig.c:
```diff
@@ -190,11 +201,11 @@ static void scan_fonts(FcConfig *config, ASS_FontProvider *provider)
     FcResult result;
 
-    if (fc->fallbacks)
+    if (!pLoc->fallbacks)
```
Whoops! Fixed.

rcombs commented:
Before landing this, we should probably tweak the `Language` header parsing and add sanitization on the field regardless. I've been experimenting with this file:
```
#include <fontconfig/fontconfig.h>
#include <stdio.h>

int main(int argc, char** argv)
{
    if (argc < 2)
        return 1;

    printf("%s -> %s\n", argv[1], FcLangNormalize(argv[1]));

    return 0;
}
```
Notably, most 3-letter codes don't get normalized to their 2-letter forms, so I'd expect 3-letter language codes wouldn't match the 2-letter values in fonts. This means that we'll need to either tell callers to pre-normalize the language codes they pass, or do it ourselves internally; the latter is probably preferable.

We should also consider what to do in the no-language case. Currently, this leaves that as status quo, which means that the results are system-locale-specific in CoreText and DirectWrite, but not in fontconfig. I think we should probably figure out a way to make this locale-specific on fontconfig as well, tbh; the user's locale is a better guess for an appropriate font language than whatever coincidental behavior we currently have.

Additionally, what should we do for cases where the caller wants to set the default language to something other than whatever the providers autodetect? Should we just tell consumers to set the track language in that case, or provide a separate API?

One potential issue with fontconfig: on a lot of systems, the default font for all CJK languages will be something like `Noto Sans CJK JP`. For instance, `fc-match :lang=ko` on my Linux system gives `NotoSansCJK-Regular.ttc: "Noto Sans CJK JP" "Regular"`. Sometimes this is due to fontconfig weirdness (I have absolutely no idea why this isn't picking the perfectly good `Noto Sans CJK KR` install that the config files point to), but some systems might actually only have one of the fonts installed, either because they're slightly smaller or because they want that language to be the default. I _think_ Harfbuzz should handle this, but we should verify that. It also means that if we decide to disable locale-variant glyph selection by default in the future (for vsfilter compat), we'll want to make sure the change only applies to attached fonts and authoring mode.

rcombs commented:
Force-pushed with:
- Language field extended to support longer strings
- Docs changed to indicate that the value should be a BCP47 tag containing only language and region

I've also tested behavior when using fonts that support multilingual glyph variants (e.g. `Noto Sans CJK JP`): Harfbuzz _does_ ensure this works correctly. If you select `Noto Sans CJK JP`, but pass language values of `zh-CN`, `zh-TW`, `zh-HK`, `ja`, or `ko`, you get the correct variants for e.g. U+8FD4:

| Lang | Glyph |
|---|---|
| zh-CN | <span lang="zh-CN">返</span> |
| zh-TW | <span lang="zh-TW">返</span> |
| zh-HK | <span lang="zh-HK">返</span> |
| ja | <span lang="ja">返</span> |
| ko | <span lang="ko">返</span> |


![](https://upload.wikimedia.org/wikipedia/commons/thumb/2/23/Source_Han_Sans_Version_Difference.svg/2560px-Source_Han_Sans_Version_Difference.svg.png)

TheOneric commented on libass/ass.c:
```diff
@@ -884,7 +884,17 @@ static int process_info_line(ASS_Track *track, char *str)
+        track->Language = strdup(p);
+        if (track->Language) {
+            size_t len = strlen(track->Language);
+            for (size_t i = len; i-- > 0;) {
+                if (track->Language[i] == ' ' ||
+                    track->Language[i] == '\t')
+                    track->Language[i] = 0;
+                else
+                    break;
+            }
```
Can't this just use `rskip_spaces` and insert a single `\0` as needed?

TheOneric commented on libass/ass_types.h:
```diff
@@ -274,7 +274,7 @@ typedef struct ass_track {
     int ScaledBorderAndShadow; // 0 or 1 (boolean)
     int Kerning; // 0 or 1 (boolean)
-    char *Language; // zero-terminated ISO-639-1 code
+    char *Language; // if non-null, zero-terminated BCP47 code including language and optional region
```
I feel like this can be misread as requiring not only the the – usually optional – region part to be included in `Language`’s BCP47 code.
Due to its importance for interoperability with ISO-639-1-only interfaces, it may also be worth to reiterate BCP47’s requirement to use 2letter codes when available.

TheOneric reviewed (commented):
It may be cleaner to split extension to BCP47 codes and `Language` usage for font fallback into separate commits.

Since `WRAP_UNICODE` was merged half a year ago, this part from my first comment here is now relevant:
> Something to consider is that different libraries we interact with support different language notations. Your commit mentions UTS#35, HarfBuzz appears to use IETF BCP 47 codes as noted in https://github.com/libass/libass/issues/404#issuecomment-648306753 and libunibreak can only handle ISO 3166-1 iirc.
> [...]
> If it lands after https://github.com/libass/libass/pull/486 the language tag must be sanitised before being passed to libunibreak.

rcombs commented:
Rebased with comments applied, and:
- Use ASS_StringView instead of a bare `const char*` in new internal API
- Have the cache items own a copy of the language tag instead of relying on the track's value never changing
- Keep runtime support for macOS versions prior to 10.9 (since we apparently care about that)
- Split into as many distinct commits as is reasonable

Anybody wanna test this on Windows?

rcombs commented:
I did some testing on fontconfig and found that the earlier version of this _sorta_ worked, but got tripped up by the `get_fallbacks` function, which gave us a long list of fallbacks to try, which would often include one for a different language before the one we were looking for. I don't think it's possible to make that list prefer locale-matching languages (or, if it _is_ possible, I think it'd be slower than what I ended up landing on without being better in any meaningful way).

The new version of this PR has a commit that vastly simplifies our fallback behavior on fontconfig (relative to the existing code, and _especially_ relative to the previous version of the PR). It now simply asks fontconfig for a match with the passed-in family, locale, and code point. This gives reasonable results for me on macOS for all tested inputs except Hong Kong Chinese (due to [this upstream configuration issue](https://gitlab.freedesktop.org/fontconfig/fontconfig/-/issues/353)), and gives reasonable results for all cases on Linux.

The main downside is that it'll be slightly slower in the case where we'd previously have simply looked up a font in the substitutions list, and the slow lookup runs once per font name that we need to fallback from. It should be much faster than the `FcFontSort` call we used to make in the `get_fallback` case, but that's offset to some extent by the results not being cached.

Also in this version:
- Added a compile-time SDK version check for the macOS language check code
- Added an `ASS_TEST_FORCE_FONTCONFIG` env var check in the `test` program to simplify testing that path on systems with other font providers

astiob commented on libass/ass_coretext.c:
```diff
@@ -270,7 +274,18 @@ static char *get_fallback(void *priv, ASS_Library *lib,
 
-    CTFontRef fb = CTFontCreateForString(font, r, CFRangeMake(0, 1));
+    CTFontRef fb = NULL;
+#if __has_builtin(__builtin_available) && defined(MAC_OS_X_VERSION_10_9)
+    if (__builtin_available(macOS 10.9, iOS 7.0, *)) {
```
It turns out this symbol isn’t actually declared in the C headers until the 10.15 SDK. As discussed on IRC, let’s declare it ourselves and hope for the best.

Further, as discussed in https://github.com/libass/libass/pull/664#discussion_r1119732310, in https://github.com/libass/libass/pull/664/commits/77e4bfb977ebd8bcde56f21da88760046a025f39 and on IRC, the following will restore compatibility with GCC and older Clang, which are viable compilers on 10.9, and make the `__builtin_available` polyfill redundant:

```suggestion
#if __MAC_OS_X_VERSION_MAX_ALLOWED >= 1090 || __IPHONE_OS_VERSION_MAX_ALLOWED >= 70000
    CTFontRef CTFontCreateForStringWithLanguage(CTFontRef currentFont, CFStringRef string, CFRange range, CFStringRef language) __OSX_AVAILABLE_STARTING(__MAC_10_9, __IPHONE_7_0);
    if ((intptr_t) &CTFontCreateForStringWithLanguage) {
```

plus `#include <Availability.h>` (or we can rely on the fact that it already happens to be included transitively, but I’d slightly prefer an explicit include),

assuming I can trust the claim in Availability.h that `__MAC_OS_X_VERSION_MAX_ALLOWED` is undefined when targeting iOS.

rcombs replied:
This introduces a `-Wpartial-availability` warning when targeting pre-10.9, which is often configured as an error.

astiob replied:
Current docs say `-Wpartial-availability` is an alias for `-Wunguarded-availability` which is the exact flag that warns about `availability`, and as you and I have both mentioned previously, it is only enabled by default in its `-Wunguarded-availability-new` which affects 10.13+ APIs, not 10.9 APIs. Our Makefile doesn’t enable it, either. This is no different from #664.

Unless perhaps you actually mean some other warning that pertains to the declaration rather than the invocation?

If you want us to support build setups that explicitly add `-Werror=unguarded-availability` to extra cflags, we need completely separate compilation paths (or a macro) for new Clang and all other compilers.

rcombs replied:
Re-pushed with a new `CHECK_AVAILABLE` macro, which should work fine for both new and old compilers.

rcombs commented on libass/ass_fontconfig.c:
```diff
@@ -202,165 +184,78 @@ static void scan_fonts(FcConfig *config, ASS_FontProvider *provider)
     .check_postscript   = check_postscript,
     .check_glyph        = check_glyph,
     .destroy_provider   = destroy,
-    .get_substitutions  = get_substitutions,
```
Ah, I've been testing this using `test/test.c`, which sets `default_family` to `"sans-serif"`. With that configuration and this patchset, the usual `serif`, `sans-serif`, and `monospace` aliases work fine for me. I get `DejaVuSerif`, `DejaVuSans`, and `DejaVuMono` for `en`, with CJK fallback on `MS-Mincho` or `MS-Gothic`; for `zh-CN`, I get `SimSun`, `ArialUnicodeMS`, and `NSimSun`; for `ja`, I get `MS-Mincho`, `MS-Gothic`, and `MS-Gothic`. This is all as expected.

With `Arial`, though, I get the same behavior you're describing here. I'm not sure what the best solution for this is; I can't see any way to implement the `get_substitutions` route in a locale-aware manner.

FlyGoat replied:
Sorry to chime in here, was bothered by mpv's font selection process for a while, so I decided to do my investigation.
`FcFontMatch` is supposed to work with alias, because it performs partial match on `FC_FAMILY`, so `ass_fontconfig` won't fail at first instance for `{\fnserif}`.
For `Arial` I guess that's because @rcombs does not have `Arial Unicode MS` installed. That's the only `Arial` family font with CJK support out there.
IMO `get_substitutions` is not really useful with `fontconfig`, given that it already handles family alias well, but perhaps performing a search without locale if first search fails can handle this situation better.

rcombs commented on libass/ass_coretext.c:
```diff
@@ -248,10 +252,10 @@ static void match_fonts(void *priv, ASS_Library *lib, ASS_FontProvider *provider
 {
-    CFStringRef name = CFStringCreateWithBytes(
-        0, (UInt8 *)family, strlen(family), kCFStringEncodingUTF8, false);
+    CFStringRef name = CFStringCreateWithCString(0, family, kCFStringEncodingUTF8);
```
I ended up removing that usage after checking and finding that I don't get reliably consistent retain counts on these objects; best not to risk it.

moi15moi commented:
Is there a reason why this PR is stalling?
It would help resolving: https://github.com/libass/libass/issues/319

TheOneric commented:
> Is there a reason why this PR is stalling?

Yes, see the open review threads: it breaks fontconfig substitutions

_(It technically also is an API change (`Language` is no longer either `NULL` or exactly three bytes long) for readers of the struct, but this should not be an issue for any remotely sane code imho.)_
